### PR TITLE
Tests & fixes off of DataSelector refactor

### PR DESCRIPTION
**Pull request by paulrosenzweig**
Branches off of #11809

- [x] Fix some interaction bugs with DataSelector
- [x] DataSelector unit tests
- [x] Schema fetching unit tests
- [x] Use new backend endpoints
- [x] Add a Cypress test or two

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/DataSelector.jsx [hunk 1]
```diff
@@ -722,9 +731,12 @@ const DatabaseSchemaPicker = ({
       isLoading,
   }));
 
-  let openSection =
-    selectedSchema &&
-    _.findIndex(databases, db => _.find(db.schemas, selectedSchema));
+  let openSection = selectedSchema
+    ? databases.findIndex(db => db.schemas.some(s => s === selectedSchema))
```
Comment (tlrobinson) [inline frontend/src/metabase/query_builder/components/DataSelector.jsx]: I think `databases.findIndex(db => db.id === selectedSchema.db.id)` should work

File: frontend/src/metabase/query_builder/components/DataSelector.jsx [hunk 2]
```diff
@@ -429,10 +438,24 @@ export default class DataSelector extends Component {
...
     const previousStep = this.getPreviousStep();
     if (previousStep) {
-      this.switchToStep(previousStep, {}, false);
+      const clearedState = this.getClearedStateForStep(previousStep);
+      this.switchToStep(previousStep, clearedState, false);
     }
   };
 
+  getClearedStateForStep(step) {
+    // TODO: do we need to clear state for other steps?
```
Comment (paulrosenzweig) [inline frontend/src/metabase/query_builder/components/DataSelector.jsx]: @tlrobinson would uncleared state be problematic for any other steps? I should probably add similar clearing code just for consistency, but it'd be nice to add some failing tests too.

File: frontend/src/metabase/query_builder/components/GuiQueryEditor.jsx [hunk 3]
```diff
@@ -331,19 +322,9 @@ export default class GuiQueryEditor extends React.Component {
...
-            setSourceTableFn={this.props.setSourceTableFn}
-            isInitiallyOpen={isInitiallyOpen}
-          />
-        ) : (
-          <span className="flex align-center px2 py2 text-bold text-grey">
-            {tableMetadata && tableMetadata.display_name}
-          </span>
-        )}
+        <span className="flex align-center px2 py2 text-bold text-grey">
+          {tableMetadata && tableMetadata.display_name}
```
Comment (tlrobinson) [inline frontend/src/metabase/query_builder/components/GuiQueryEditor.jsx]: Not a big deal but if you feel like it we could use the newer style:

```js
const table = this.props.query.table();

...

{table && table.displayName()}
```

File: frontend/src/metabase/query_builder/components/dataref/DatabasePane.jsx [hunk 4]
```diff
@@ -1,16 +1,40 @@
...
-    props.database.schemas.length > 1
-      ? DatabaseSchemasPane
-      : DatabaseTablesPane;
-  return <Component {...props} />;
-};
+@Databases.load({
+  id: (state, { database }) => database && database.id,
+  wrapped: true,
+})
+class DatabasePane extends Component {
```
Comment (tlrobinson) [inline frontend/src/metabase/query_builder/components/dataref/DatabasePane.jsx]: Also not a big deal but I prefer `React.Component`

File: frontend/test/metabase/query_builder/components/DataSelector.unit.spec.js [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+  state as fixtureData,
+} from "__support__/sample_dataset_fixture";
+
+import { UnconnectedDataSelector as DataSelector } from "metabase/query_builder/components/DataSelector";
+
+describe("DataSelector", () => {
+  let originalTimeout;
+  beforeEach(() => {
+    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
+    jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;
```
Comment (tlrobinson) [inline frontend/test/metabase/query_builder/components/DataSelector.unit.spec.js]: Why so long?

File: frontend/test/metabase/entities/databases.unit.spec.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import mock from "xhr-mock";
```
Comment (tlrobinson) [inline frontend/test/metabase/entities/databases.unit.spec.js]: Just FYI (no need to change this), all of our similar tests that mock XHR are considered "integration" tests (`*.integ.spec.js`) but honestly the distinction isn't particularly useful.

I think `*.integ.spec.js` should actually use real API calls or at least snapshots of real API calls.

File: frontend/src/metabase/query_builder/components/DataSelector.jsx [hunk 7]
```diff
@@ -683,12 +705,12 @@ const SchemaPicker = ({
...
-        id="DatabaseSchemaPicker"
-        key="databaseSchemaPicker"
+        id="SchemaPicker"
+        key="schemaPicker"
         className="text-brand"
         sections={sections}
         searchable
-        onChange={onChangeSchema}
+        onChange={item => onChangeSchema(item.schema)}
         itemIsSelected={schema => schema === selectedSchema}
```
Comment (tlrobinson) [inline frontend/src/metabase/query_builder/components/DataSelector.jsx]: Should this be `itemIsSelected={item => item.schema === selectedSchema}` ?

**Review by tlrobinson — COMMENTED** [review]
A few little things but looks great.

File: frontend/test/metabase/query_builder/components/DataSelector.unit.spec.js (hunk 5 above)
Comment (paulrosenzweig) [reply frontend/test/metabase/query_builder/components/DataSelector.unit.spec.js]: This one was weird. I still don't know the cause. Without the increased timeout, this suite passes alone but consistently fails when running all of them. Googling indicated that the timeout error can be due to conflict between Promise libraries. I tried running different subsets of suites and there is some "tipping point", but it's not consistently any one file that clashes.

I have no idea what would take more than five seconds, but bumping it seemed to make it consistently pass. 🤷‍♂

File: frontend/src/metabase/new_query/containers/NewQueryOptions.jsx [hunk 8]
```diff
@@ -28,7 +28,8 @@ const mapStateToProps = state => ({
 });
 
 const mapDispatchToProps = {
-  prefetchMetadata: () => Database.actions.fetchList({ include_tables: true }),
+  prefetchMetadata: () =>
```
Comment (tlrobinson) [inline frontend/src/metabase/new_query/containers/NewQueryOptions.jsx]: Have you verified if this actually avoids needing to fetch individuals dbs/schemas?

**Review by tlrobinson — APPROVED** [review]
One issue I noticed is we're showing the "Saved Questions" database even on brand new instances where there are no saved questions. We could probably fix that by not returning the db from the endpoint if there are no saved questions.

Otherwise this looks good.

File: frontend/src/metabase/new_query/containers/NewQueryOptions.jsx (hunk 8 above)
Comment (paulrosenzweig) [reply frontend/src/metabase/new_query/containers/NewQueryOptions.jsx]: Yup! I manually verified and added [a test](https://github.com/metabase/metabase/blob/c55bf11d9a74445450ab56a98ef04a2f2409e0c8/frontend/test/metabase/scenarios/question/query_builder_loading.cy.spec.js), which I'm "meh" on the utility of.

It doesn't avoid fetching the saved questions db, but I think that's better. We only do that big load if the user actually wants to see it.

**Pull request merged**